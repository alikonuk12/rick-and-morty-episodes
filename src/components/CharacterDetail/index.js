import React, { useEffect, useState } from 'react';
import styles from './index.module.scss';
import { useParams } from 'react-router-dom';
import { getById } from '../../api/character';
import { getByLink } from '../../api/episode';
import { EpisodeCard, Spinner } from '../../reusables';

const CharacterDetail = () => {
    const [char, setChar] = useState(undefined);
    const [episodes, setEpisodes] = useState([]);
    const [isReady, setIsReady] = useState(false);

    const { character } = useParams();

    const prepare = async () => {
        if (character) {
            let arr = [];
            const result = await getById(character);
            setChar(result);
            for (let i = 0; i < result.episode.length; i++) {
                const ep = await getByLink(result.episode[i]);
                arr.push(ep);
            }
            setEpisodes(arr);
            setIsReady(true);
        }
    }

    useEffect(() => {
        prepare();
    }, []);

    return (
        <div>
            {
                char &&
                <>
                    <div className={styles.infoContainer}>
                        <div>
                            <img className={styles.image} src={char.image} alt={char.name} />
                        </div>
                        <div>
                            <div className={styles.name}>{char.name}</div>
                            <div className={styles.statusContainer}>
                                {<div className={char.status === 'Alive' ? styles.greenDot : char.status === 'Dead' ? styles.redDot : styles.lightgrayDot} />}
                                <div className={styles.status}>{char.status} - {char.species}</div>
                            </div>
                            <div className={styles.subContainer}>
                                <div className={styles.subTitle}>Type: </div>
                                <div className={styles.type}>{char.type}</div>
                            </div>
                            <div className={styles.subContainer}>
                                <div className={styles.subTitle}>Gender: </div>
                                <div className={styles.gender}>{char.gender}</div>
                            </div>
                            <div className={styles.titleContainer}>
                                <div className={styles.subTitle}>Last known location:</div>
                                <div className={styles.location}>{char.location.name}</div>
                            </div>
                            <div className={styles.titleContainer}>
                                <div className={styles.subTitle}>First seen in:</div>
                                <div className={styles.origin}>{char.origin.name}</div>
                            </div>
                        </div>
                    </div>
                    <div className={styles.episodeContainer}>
                        <div className={styles.episodeHeader}>Episodes with {char.name}</div>
                        {
                            isReady ?
                                <div className={styles.episodes}>
                                    {
                                        episodes.map((value, index) => {
                                            return <EpisodeCard key={index} episode={value} />
                                        })
                                    }
                                </div> :
                                <div className={styles.spinnerContainer}>
                                    <Spinner />
                                </div>
                        }
                    </div>
                </>
            }
        </div>
    );
}

export default CharacterDetail;