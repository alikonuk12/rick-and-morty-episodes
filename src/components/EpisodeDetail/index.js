import React, { useState, useEffect } from 'react';
import styles from './index.module.scss';
import { useParams } from 'react-router-dom';
import { Button, CharacterCard, Filter, Spinner, Dropdown } from '../../reusables';
import { getById } from '../../api/episode';
import { getByLink } from '../../api/character';

const EpisodeDetail = () => {
    const [characters, setCharacters] = useState([]);
    const [search, setSearch] = useState('');
    const [sort, setSort] = useState(true);
    const [isReady, setIsReady] = useState(false);
    const [status, setStatus] = useState(undefined);
    const [gender, setGender] = useState(undefined);

    const statusArr = ['Select Status', 'Alive', 'Dead', 'unknown'];
    const genderArr = ['Select Gender', 'Female', 'Male', 'Genderless', 'unknown'];

    const { episode } = useParams();

    const handleChangeSearch = e => setSearch(e.target.value);
    const handleClickSort = () => setSort(!sort);

    const handleChangeStatus = e => setStatus(e.target.value);
    const handleChangeGender = e => setGender(e.target.value);

    const decrease = (a, b) => {
        if (a.name < b.name) return -1;
        if (b.name < a.name) return 1;
        return 0;
    }

    const increase = (a, b) => {
        if (a.name < b.name) return 1;
        if (b.name < a.name) return -1;
        return 0;
    }

    const prepare = async () => {
        if (episode) {
            let arr = [];
            const result = await getById(episode);
            for (let i = 0; i < result.characters.length; i++) {
                const char = await getByLink(result.characters[i]);
                arr = arr.concat(char);
            }
            setCharacters(arr);
            setIsReady(true);
        }
    }

    useEffect(() => {
        prepare();
    }, []);

    return (
        <div className={styles.container}>
            <div className={styles.filterButtonContainer}>
                <Filter value={search} onChange={handleChangeSearch} />
                <Dropdown value={status} list={statusArr} onChange={handleChangeStatus} />
                <Dropdown value={gender} list={genderArr} onChange={handleChangeGender} />
                <Button text={sort ? 'A -> Z' : 'Z -> A'} onClick={handleClickSort} />
            </div>
            <div className={styles.episodesContainer}>
                {
                    isReady && characters.length > 0 ?
                        <>
                            {
                                characters
                                    .filter(el => el.name.toLowerCase().startsWith(search.toLowerCase()))
                                    .filter(el => {
                                        if (status && !status.startsWith('Select')) {
                                            return el.status === status;
                                        } else {
                                            return el;
                                        }
                                    })
                                    .filter(el => {
                                        if (gender && !gender.startsWith('Select')) {
                                            return el.gender === gender;
                                        } else {
                                            return el;
                                        }
                                    })
                                    .sort(sort ? decrease : increase)
                                    .map((value, index) => {
                                        return <CharacterCard key={index} character={value} />
                                    })
                            }
                        </>
                        :
                        <div className={styles.spinnerContainer}>
                            <Spinner />
                        </div>

                }
            </div>
        </div>
    );
}

export default EpisodeDetail;