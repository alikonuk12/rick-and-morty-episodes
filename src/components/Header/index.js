import React from 'react';
import styles from './index.module.scss';
import { NavLink } from 'react-router-dom';
import Banner from '../../assets/img/banner.png';

const Header = () => {
    return (
        <header>
            <div className={styles.header}>
                <NavLink
                    exact='true'
                    to="/" className={navData => navData.isActive ? styles.headerLinkActive : styles.headerLink}
                >
                    <img src={Banner} className={styles.img} alt='banner' />
                </NavLink>
                <NavLink
                    exact='true'
                    to="/"
                    className={navData => navData.isActive ? styles.headerLinkActive : styles.headerLink}
                >
                    Home
                </NavLink>
            </div>
        </header>
    )
}

export default Header;