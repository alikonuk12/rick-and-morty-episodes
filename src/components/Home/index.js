import React, { useEffect, useState } from 'react';
import styles from './index.module.scss';
import { get } from '../../api/episode';
import { EpisodeCard } from '../../reusables';

const Home = () => {
    const [episodes, setEpisodes] = useState([]);

    const prepare = async () => {
        let arr = [];
        for (let i = 1; i < 4; i++) {
            const { results } = await get(i);
            arr = arr.concat(results);
        }
        setEpisodes(arr);
    }

    useEffect(() => {
        prepare();
    }, []);

    return (
        <div className={styles.container}>
            {
                episodes.length > 0 &&
                episodes.map((value, index) => {
                    return <EpisodeCard key={index} episode={value} />
                })
            }
        </div>
    );
}

export default Home;