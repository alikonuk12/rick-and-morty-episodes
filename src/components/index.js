import Header from './Header';
import Home from './Home';
import EpisodeDetail from './EpisodeDetail';
import CharacterDetail from './CharacterDetail';

export { Header, Home, EpisodeDetail, CharacterDetail };