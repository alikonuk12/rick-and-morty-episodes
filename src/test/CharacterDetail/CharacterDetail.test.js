import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { CharacterDetail } from '../../components';

Enzyme.configure({ adapter: new Adapter() });

describe('CharacterDetail', () => {
    it('rendering character detail', () => {
        const component = shallow(<CharacterDetail />);
        expect(component).not.toBe(null);
        expect(component.getElements()).toMatchSnapshot();
    });
});