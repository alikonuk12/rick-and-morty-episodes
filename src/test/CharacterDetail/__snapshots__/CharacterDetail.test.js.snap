// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterDetail rendering character detail 1`] = `
Array [
  <div />,
]
`;
